[RSDK-10933] replace most function pointers in plannerOptions with enums

gvaradarajan opened the pull request:
This is the first of two PRs removing function pointers from the `plannerOptions` struct as part of an ongoing effort to make `PlanRequest` serializable.

The remaining function pointers are in the `pathMetric` attribute and the `ConstraintHandlers` struct

biotinker commented:
I think I would prefer a slightly different approach.

It shouldn't be a problem for private, internal structs to contain function pointers; and when doing testing/debugging on `motionplan`, it's not that uncommon for me to manually define my own version of these functions.

I think part of the problem we have right now is that `plannerOptions` is a single struct used both for json unmarshalling, and also for storing/passing much more complicated things like function pointers.

How would we feel about splitting this functionality in half, having a public struct which unmarshals json, is populated only with atomics (including enums etc), which is parsed into an internal struct which contains function pointers among other things?

gvaradarajan commented:
> It shouldn't be a problem for private, internal structs to contain function pointers; and when doing testing/debugging on `motionplan`, it's not that uncommon for me to manually define my own version of these functions.
> 

You can still do that by editing the logic in the functions that receive the enums. `plannerOptions` will eventually not be private and I disagree with the idea that public API design should be organized around this need.

> I think part of the problem we have right now is that `plannerOptions` is a single struct used both for json unmarshalling, and also for storing/passing much more complicated things like function pointers.
> 
> How would we feel about splitting this functionality in half, having a public struct which unmarshals json, is populated only with atomics (including enums etc), which is parsed into an internal struct which contains function pointers among other things?

My plan going forward is to move the function pointers for things that are not as easily parametrized (like those in `ConstraintHandler`) into the `planner` struct and out of an options object. I think it works better when options contain information and not functions. These enums are just laying the groundwork for that move (because I didn't want to drop a 5000 line PR).

gvaradarajan commented:
> > You can still do that by editing the logic in the functions that receive the enums. plannerOptions will eventually not be private and I disagree with the idea that public API design should be organized around this need.
> 
> I also disagree with the idea that public API design should be organized around this need. :)
> 
> I don't think I suggested any specific changes to the (current or future) public API whatsoever; we're on the same page there. What we're discussing is the difference between the internal, private organization being enums -> an abstraction function that returns a function and calling that function every time we need the function pointer, vs calling it once up front and storing it somewhere- such as on the `planner`, as you suggested for the constraint handler.

Let me see if I've understood. So, for example, you would like the result of `getScoringFunction` to be stored somewhere rather than repeatedle called?

biotinker commented:
> Let me see if I've understood. So, for example, you would like the result of getScoringFunction to be stored somewhere rather than repeatedly called?

Yes. I agree that `plannerOptions` should be made public, and that that public struct should be serializable, and enums are a great way to do that.

What I don't love is the repeated lookups, including in the middle of planning algorithms. Things that will definitely not change and can be cached ahead of time, IMO generally should be unless there's some specific reason not to.

What that looks like could be these functions getting added onto `planner`, or some new private struct, or something else.

gvaradarajan commented:
@biotinker I've brought back caching some of the functions but moved them to the `planner` struct. Please let me know whether I've sufficiently addressed your concerns

gvaradarajan commented on motionplan/rrt.go:
```diff
@@ -96,8 +96,10 @@ func initRRTSolutions(ctx context.Context, wp atomicWaypoint) *rrtSolution {
 		return rrt
 	}
 
+	configDistMetric := ik.GetConfigurationDistanceFunc(wp.mp.opt().ConfigurationDistanceMetric)
```
`wp.mp` is just an instance of the `motionPlanner` interface, it doesn't know for sure about the struct member. Did you want it to be returnable by the interface?

biotinker replied:
No- never mind

biotinker reviewed (approved):
LGTM
